### PR TITLE
[Draft] wNAF multiplication for BN256 curve

### PR DESCRIPTION
**Pull request by ZamDimon**
# What ❔

This pull request implements multiplication by a scalar for the `BN256` curve. This PR adds:
- WNAF-based implementation, mostly based on [this implementation](https://github.com/matter-labs/era-zkevm_circuits/blob/main/src/ecrecover/new_optimized.rs#L319).
- `.sage` files for finding: 
  - $\beta$ and $\lambda$ parameters for useful endomorphism;
  - short vectors $(a_1,b_1)$ and $(a_2,b_2)$ for effective scalar decomposition

## Why ❔

Currently, multiplication by a scalar is not implemented for BN256 and therefore one needs to add support for such an operation.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: src/gadgets/curves/bn256/decomp_table.rs
Comment (jules) [inline src/gadgets/curves/bn256/decomp_table.rs:1]: As discussed, this can be deleted

File: src/gadgets/curves/bn256/naf_abs_div2_table.rs
Comment (jules) [inline src/gadgets/curves/bn256/naf_abs_div2_table.rs:1]: Can also be deleted I believe

File: src/gadgets/curves/bn256/mul.rs [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+            )
+        };
+    }
+
+    UInt256 { inner: limbs }
+}
+
+/// Converts the specified `BN256ScalarNNField` element to a
+/// WNAF representation using an array of `UInt8<F>` elements.
+fn convert_to_wnaf<F: SmallField, CS: ConstraintSystem<F>>(
```
Comment (jules) [inline src/gadgets/curves/bn256/mul.rs:146]: All wNAF functionality can be removed. We keep the GLV decomposition but instead use Alex's fixed-width 4-bit mul circuit found [here](https://github.com/matter-labs/era-zkevm_circuits/blob/v1.4.1/src/ecrecover/new_optimized.rs#L328). Note that the code here deals with potential 129 bit decomposed scalars, whereas bn254 will only occupy max 127 bits. Thus, there should be some extra constraints we can drop as the scalars fit in a smaller size.

**Review by jules — CHANGES_REQUESTED** [review]
Everything looks great - as discussed, let's move to the cheaper mul circuit and we should be done here :)

**Comment by ZamDimon** [comment]
Hey @jules, enormous thanks for the review! Updated the wnaf-based scalar multiplication to a fixed-width 4-bit multiplication, as requested, according to implementation [here](https://github.com/matter-labs/era-zkevm_circuits/blob/v1.4.1/src/ecrecover/new_optimized.rs#L328).

After your comment, I am now aware of the extra circuit complexity due to dealing with the potential 129-bit decomposed scalar. Still, I suggest staying with the current implementation since it is well-tested. Then, after we successfully integrate the `ecMul` in the more high-level repositories and make sure it works, we will surely cut off the constraints.

**Comment by ZamDimon** [comment]
BN254-specific functions will be transferred to [`era-zkevm_circuits`](https://github.com/matter-labs/era-zkevm_circuits/pull/32). For helper gadgets, see #38. Thus, we close this pull request.

**Pull request closed without merging**